qiskit/transpiler/target.py

```diff
@@ -0,0 +1,652 @@
+        "num_qubits",
+        "_gate_map",
+        "_gate_name_map",
+        "_qarg_gate_map",
+        "description",
+        "_coupling_graph",
+        "_instruction_durations",
+        "_instruction_schedule_map",
+        "dt",
+        "granularity",
+        "min_length",
+        "pulse_alignment",
+        "aquire_alignment",
+    )
+
+    def __init__(
+        self,
+        description=None,
+        num_qubits=0,
+        dt=None,
+        granularity=1,
+        min_length=1,
+        pulse_alignment=1,
+        aquire_alignment=1,
+    ):
+        """
+        Create a new Target object
+
+        Args:
+            description (str): An optional string to describe the Target.
+            num_qubits (int): An optional int to specify the number of qubits
+                the backend target has. If not set it will beimplicitly set
+                based on the qargs when :meth:`~qiskit.Target.add_instruction`
+                is called. Note this must be set if the backend target is for a
+                noiseless simulator that doesn't have constraints on the
+                instructions so the transpiler knows how many qubits are
+                available.
+            dt (float): The system time resolution of input signals in seconds
+            granularity (int): An integer value representing minimum pulse gate
+                resolution in units of ``dt``. A user-defined pulse gate should
+                have duration of a multiple of this granularity value.
+            min_length (int): An integer value representing minimum pulse gate
+                length in units of ``dt``. A user-defined pulse gate should be
+                longer than this length.
+            pulse_alignment (int): An integer value representing a time
+                resolution of gate instruction starting time. Gate instruction
+                should start at time which is a multiple of the alignment
+                value.
+            acquire_alignment (int): An integer value representing a time
+                resolution of measure instruction starting time. Measure
+                instruction should start at time which is a multiple of the
+                alignment value.
+        """
+        self.num_qubits = num_qubits
+        # A mapping of gate name -> gate instance
+        self._gate_name_map = {}
+        # A nested mapping of gate name -> qargs -> properties
+        self._gate_map = {}
+        # A mapping of qarg -> set(gate name)
+        self._qarg_gate_map = defaultdict(set)
+        self.dt = dt
+        self.description = description
+        self._coupling_graph = None
+        self._instruction_durations = None
+        self._instruction_schedule_map = None
+        self.granularity = granularity
+        self.min_length = min_length
+        self.pulse_alignment = pulse_alignment
+        self.aquire_alignment = aquire_alignment
+
+    def add_instruction(self, instruction, properties=None, name=None):
```

Comment:
Having `qarg` be part of the `properties` dict feels somewhat unnatural here since other methods (e.g. `update_instruction_properties`) promote `qargs` as a top level argument, though this is probably very deeply baked into the design at this point.

Reply:
It was separate originally see the discussion on this 2 months ago in: https://github.com/Qiskit/qiskit-terra/pull/5885#discussion_r714861435 we updated it to make it easier to avoid situations where `qargs != properties.keys()`.